server/sonar-web/src/main/js/apps/marketplace/App.tsx

```diff
@@ -0,0 +1,274 @@
+import Search from './Search';
+import {
+  getAvailablePlugins,
+  getInstalledPluginsWithUpdates,
+  getPendingPlugins,
+  getPluginUpdates,
+  Plugin,
+  PluginPending
+} from '../../api/plugins';
+import { Edition, EditionStatus, getEditionsList, getEditionStatus } from '../../api/marketplace';
+import { RawQuery } from '../../helpers/query';
+import { translate } from '../../helpers/l10n';
+import {
+  getEditionsForLastVersion,
+  getEditionsForVersion,
+  filterPlugins,
+  parseQuery,
+  Query,
+  serializeQuery
+} from './utils';
+
+export interface Props {
+  editionsUrl: string;
+  location: { pathname: string; query: RawQuery };
+  sonarqubeVersion: string;
+  standaloneMode: boolean;
+  updateCenterActive: boolean;
+}
+
+interface State {
+  editions?: Edition[];
+  editionsReadOnly: boolean;
+  editionStatus?: EditionStatus;
+  loadingEditions: boolean;
+  loadingPlugins: boolean;
+  pending: {
+    installing: PluginPending[];
+    updating: PluginPending[];
+    removing: PluginPending[];
+  };
+  plugins: Plugin[];
+}
+
+export default class App extends React.PureComponent<Props, State> {
+  mounted: boolean;
+  timer?: NodeJS.Timer;
+
+  static contextTypes = {
+    router: PropTypes.object.isRequired
+  };
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      editionsReadOnly: false,
+      loadingEditions: true,
+      loadingPlugins: true,
+      pending: {
+        installing: [],
+        updating: [],
+        removing: []
+      },
+      plugins: []
+    };
+  }
+
+  componentDidMount() {
+    this.mounted = true;
+    this.fetchEditions();
+    this.fetchPendingPlugins();
+    this.fetchEditionStatus();
+    this.fetchQueryPlugins();
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.location.query.filter !== this.props.location.query.filter) {
+      this.fetchQueryPlugins();
+    }
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+  }
+
+  fetchQueryPlugins = () => {
+    const query = parseQuery(this.props.location.query);
+    if (query.filter === 'updates') {
+      this.fetchUpdatesOnly();
+    } else {
+      this.fetchAllPlugins();
+    }
+  };
+
+  fetchAllPlugins = () => {
+    this.setState({ loadingPlugins: true });
+    Promise.all([getInstalledPluginsWithUpdates(), getAvailablePlugins()]).then(
+      ([installed, available]) => {
+        if (this.mounted) {
+          this.setState({
+            loadingPlugins: false,
+            plugins: sortBy(uniqBy([...installed, ...available.plugins], 'key'), 'name')
+          });
+        }
+      },
+      () => {
+        if (this.mounted) {
+          this.setState({ loadingPlugins: false });
+        }
+      }
+    );
+  };
+
+  fetchUpdatesOnly = () => {
+    this.setState({ loadingPlugins: true });
+    getPluginUpdates().then(
+      plugins => {
+        if (this.mounted) {
+          this.setState({ loadingPlugins: false, plugins });
+        }
+      },
+      () => {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Update this function so that its implementation is not identical to the one on line 134. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4144)